### ISSUE TITLE
Use FunctionWrappers.jl for FinSet and FinRel

### ISSUE THREAD

**Issue by epatters** [AlgebraicJulia/Catlab.jl#225]
Specifically, in the concrete types `FinFunctionCallable` and `FinRelationCallable`.

**Comment by epatters** [comment]
@olynch, this makes sense, right? If so, would you be willing to take care of it?

**Comment by olynch** [comment]
Yes, I can do this.

**Comment by epatters** [comment]
Awesome, thanks!

**Comment by olynch** [comment]
I think that in order to do this, FinRel/FinSet has to be parametric in not only the type of storage but also the type of element. A `FunctionWrapper{Int, Tuple{Int}}` doesn't always work, because if `S` is, say, `Set{Symbol}`, it won't work. Should I do this refactor?

**Comment by epatters** [comment]
I see your point. Yes, please do. I guess we should have `FinSet{S,T}` where `S` is the underlying set type and `T` is the element type. This should be backwards compatible (I'd feel bad about breaking this interface again) and still allow convenient dispatch through `FinSet{Int}` and `FinSet{<:AbstractSet}`. What do you think?

**Comment by olynch** [comment]
Sounds good to me!

**Comment by olynch** [comment]
It seems to me that this will end up with the category `FinSet` being the disjoint union of categories `FinSet{S,T}`. This is not necessarily a bad thing, just am pointing it out in case it is objectionable.

**Comment by epatters** [comment]
Hmm, so couldn't have, say, a morphism from `FinSet{Set{Symbol}}` to `FinSet{Set{Int}}` using `FunctionWrapper{Int,Tuple{Symbol}}`?

**Comment by olynch** [comment]
Well, I would have to make

```
@auto_hash_equals struct FinFunctionCallable{S1,T1,S2,T2} <: FinFunction{S1,T1,S2,T2}
  func::FunctionWrapper{T1,Tuple{T2}} # Usually `Function` but can be any Julia callable.
  dom::FinSet{S1,T1}
  codom::FinSet{S2,T2}
end
```

**Comment by epatters** [comment]
OK, for now let's just assume that `FinFunction`s are always between the same `FinSet{S,T}`. We can revisit this if it becomes a problem in the future. (At the moment, I don't think anybody is using anything besides `FinSet{Int}`!)

**Comment by olynch** [comment]
Sounds good to me!

**Comment by epatters** [comment]
Resolved by #239.
